### ISSUE TITLE
size of actual data in the metadata

### ISSUE THREAD

**Issue by iwanbk** [zero-os/0-stor#500]
To be used as minio backend, we need to know the size of actual data in the metadata.

From my test, both meta.Size and chunk.Size is the size of processed (compressed & encrypted) data.

As far as i can see, it is a mandatory feature @zaibon

**Comment by iwanbk** [comment]
> As far as i can see, it is a mandatory feature @zaibon

When downloading data, the client do these:
- get size of the data -> this is why we need this feature
- download the data using `content-length` header with value = size of the data. I guess it will use `offset` and `length` if the size is too big, as described at #499

**Comment by GlenDC** [comment]
Can you link me to the minio docs that describe these mandatory features you're talking about?

> From my test, both meta.Size and chunk.Size is the size of processed (compressed & encrypted) data.

That is correct, however, if you would like, we could implement the actual _total_ data in the metadata, it shouldn't be that hard.

**Comment by iwanbk** [comment]
See `GetObjectInfo`  of https://godoc.org/github.com/minio/minio/cmd#ObjectLayer.
It return https://godoc.org/github.com/minio/minio/cmd#ObjectInfo which has `Size` field.

Which is needed when downloading object as described at https://github.com/zero-os/0-stor/issues/500#issuecomment-362213123

**Comment by iwanbk** [comment]
After did a quick read on the code, it seems this is the way to go

Change `Pipeline.Write` interface to `Write(r io.Reader) ([]metatypes.Chunk, int, error) ` where `int` is the size of actual data.

And then modify these func to also return actual data
```go
(sop *SingleObjectPipeline) Write(r io.Reader) ([]metatypes.Chunk, error)
(asp *AsyncSplitterPipeline) Write(r io.Reader) ([]metatypes.Chunk, error)
```

`(sop *SingleObjectPipeline) Write(r io.Reader)` is very straightforward to modify.

For `(asp *AsyncSplitterPipeline) Write(r io.Reader) ([]metatypes.Chunk, error)`, we can modify this part

https://github.com/zero-os/0-stor/blob/1fed4cb32f51c31c5b9f99581eebdb0b8fe44ff9/client/datastor/pipeline/async_splitter.go#L157-L184

make it to also count the total data.

**Comment by iwanbk** [comment]
what do you think @GlenDC ?

**Comment by GlenDC** [comment]
I don't think I agree, but I do like the direction you're going. Will take a look at it later this week.

**Comment by iwanbk** [comment]
>  Will take a look at it later this week.

Any update about this?

**Comment by zaibon** [comment]
I think we should store the actual size of the original file in the metadata as well, at the end this is a usefull information to have anyhow.
So I would just add a field  in the `medatatypes.Metadata` and then update the required code to fill this afftribute

**Comment by GlenDC** [comment]
> think we should store the actual size of the original file in the metadata as well, at the end this is a usefull information to have anyhow.

Yes I agree. We could add the property and call it `Size`, and rename the other one (the one that is currently called Size`) to `StorageSize`.

> ...
> To count the size from a reader we can use a trick like this: https://play.golang.org/p/65fZhEoWq2Z
> ...

I agree, sounds like a simple and solid solution. It's simple enough, and keeps the logic tied to the (master) client (package).

**Comment by GlenDC** [comment]
Closed in 7a384be7a8d91c7f39c981a565a6b096543f08ef.
